### ISSUE TITLE
Template inheritance allows violating parent type parameter constraints

### ISSUE THREAD

**Issue by weirdan** [vimeo/psalm#1244]
https://getpsalm.org/r/0742785937
```php
<?php

  /**
   * @template T as object
   */
  class Base {
     /** @param T $_o */
     public function __construct($_o) {}
     /**
      * @return T
      * @psalm-suppress InvalidReturnType
      */
     public function t() {}
  }

  // specializing by instantiation

//  new Base(1); // reported, as expected

  // specializing by inheritance

  /** @template-extends Base<int> */
  class SpecializedByInheritance extends Base {}  // not reported, but should be, as int is not an object

  // specializing by instantiation through inheritance
  
  /**
   * @template T
   * @template-extends Base<T>
   */
  class PassThroughGeneric extends Base {}
  new PassThroughGeneric(1); // not reported, but should be, as Base::__construct is called with int instead of object.
                             // However it's probably more useful to require T to be at least as constrained as Base's T
  
 
  /**
   * @template T
   * @param T $p
   * @return PassThroughGeneric<T>
   */
   function make($p) { return new PassThroughGeneric($p); }

   make(1); // not reported, but generates PassThroughGeneric<int>, which should not be possible as per above.
            // Again, it's probably should be reported where we declare PassThroughGeneric with a non-sufficiently 
            // constrained type parameter (the @return annotation)
```

See inline comments

**Comment by weirdan** [comment]
@muglug great work :1st_place_medal: 

What about `@return`, `@var` and `@param` annotations - they still appear to allow the type var constraints to be violated. Shall I report this separately?

**Comment by muglug** [comment]
Yes please!
